File: ml-ranking-system/generate_training_data.py
"""
Generate training data from MongoDB for ML model retraining
"""

import pandas as pd
import numpy as np
from pymongo import MongoClient
from pathlib import Path


def generate_training_data():
    """Connect to MongoDB and generate training data CSV"""
    
    # Connect to MongoDB
    client = MongoClient('mongodb://localhost:27017/')
    db = client['GmartDb']
    products_collection = db['products']
    brands_collection = db['brandreputations']
    
    print("Fetching products from MongoDB...")
    products = list(products_collection.find({}))
    print(f"Found {len(products)} products")
    
    # Build brand reputation map
    print("Fetching brand reputations...")
    brand_reps = list(brands_collection.find({}))
    brand_rep_map = {}
    for br in brand_reps:
        brand_rep_map[br['brand']] = br.get('reputationScore', 5)
    print(f"Fetched {len(brand_rep_map)} brand reputations")
    
    # Generate training data
    rows = []
    for product in products:
        brand = product.get('brand', 'Unknown')
        price = product.get('price', 1000)
        
        # Create random budget ranges for training
        budget_min = max(0, price - 500)
        budget_max = price + 500
        
        # Create quality label based on product attributes
        rating = product.get('averageRating', 3) / 5.0  # Normalize to 0-1
        reviews = min(product.get('totalReviews', 0), 100) / 100.0  # Normalize
        brand_score = brand_rep_map.get(brand, 5) / 10.0  # Normalize to 0-1
        price_score = max(0, 1 - (price / 10000))  # Lower price = higher score
        
        # Calculate quality label (0-4 scale)
        quality = rating * 0.3 + reviews * 0.25 + brand_score * 0.25 + price_score * 0.2
        label = min(4, int(quality * 5))
        
        row = {
            'price': price,
            'discountPercentage': product.get('discount', 0),
            'averageRating': product.get('averageRating', 3),
            'brandReputationScore': brand_rep_map.get(brand, 5),
            'totalReviews': product.get('totalReviews', 0),
            'budgetMin': budget_min,
            'budgetMax': budget_max,
            'label': label
        }
        rows.append(row)
    
    # Create DataFrame and save CSV
    df = pd.DataFrame(rows)
    output_dir = Path('data')
    output_dir.mkdir(exist_ok=True)
    output_path = output_dir / 'training_data.csv'
    
    df.to_csv(output_path, index=False)
    print(f"\nTraining data saved to {output_path}")
    print(f"Total samples: {len(df)}")
    print(f"Label distribution:\n{df['label'].value_counts().sort_index()}")
    
    client.close()


if __name__ == '__main__':
    generate_training_data()

File: ml-ranking-system/generate_training_data.js
/**
 * Generate training data from MongoDB for ML model retraining
 * Outputs CSV with all required features including brandReputationScore
 */

const mongoose = require('mongoose');
const fs = require('fs');
const path = require('path');

// Connect to MongoDB
mongoose.connect('mongodb://localhost:27017/GmartDb', {
  useNewUrlParser: true,
  useUnifiedTopology: true
});

// Define Product schema locally to match backend
const productSchema = new mongoose.Schema({
  name: String,
  category: String,
  price: Number,
  description: String,
  image: String,
  brand: String,
  variant: {
    color: String,
    capacity: String
  },
  discount: Number,
  rating: Number,
  reviews: [Object],
  productReviews: [Object],
  averageRating: Number,
  totalReviews: Number
});

const brandReputationSchema = new mongoose.Schema({
  brand: String,
  reputationScore: Number,
  meta: {
    colorPopularity: Object,
    avgRating: Number
  }
});

const Product = mongoose.model('Product', productSchema);
const BrandReputation = mongoose.model('BrandReputation', brandReputationSchema);

async function generateTrainingData() {
  try {
    console.log('Fetching products and brand reputations...');
    
    // Fetch all products
    const products = await Product.find({}).lean();
    console.log(`Found ${products.length} products`);
    
    // Fetch brand reputations
    const brandReps = await BrandReputation.find({}).lean();
    const brandRepMap = {};
    brandReps.forEach(br => {
      brandRepMap[br.brand] = br.reputationScore || 5;
    });
    console.log(`Fetched ${Object.keys(brandRepMap).length} brand reputations`);
    
    // Generate training data rows
    const rows = [];
    
    // Add header
    rows.push('price,discountPercentage,averageRating,brandReputationScore,totalReviews,budgetMin,budgetMax,label');
    
    products.forEach((product, idx) => {
      // Assign random budget ranges (training samples)
      const budgetMin = Math.max(0, product.price - 500);
      const budgetMax = product.price + 500;
      
      // Assign random quality label (0-4 = ranks from worst to best)
      // Higher rating + higher reviews + higher brand rep + lower price = higher quality
      const ratingScore = (product.averageRating || 3) / 5;  // 0-1
      const reviewScore = Math.min(product.totalReviews || 0, 100) / 100;  // 0-1
      const brandScore = (brandRepMap[product.brand] || 5) / 10;  // 0-1
      const priceScore = Math.max(0, 1 - (product.price / 10000));  // 0-1, lower price better
      
      // Weighted quality label
      const qualityScore = (ratingScore * 0.3 + reviewScore * 0.25 + brandScore * 0.25 + priceScore * 0.2);
      const label = Math.min(4, Math.floor(qualityScore * 5));  // 0-4 scale
      
      const row = [
        product.price || 1000,
        product.discount || 0,
        product.averageRating || 3,
        brandRepMap[product.brand] || 5,
        product.totalReviews || 0,
        budgetMin,
        budgetMax,
        label
      ].join(',');
      
      rows.push(row);
    });
    
    // Write CSV file
    const outputPath = path.join(__dirname, 'data', 'training_data.csv');
    const dir = path.dirname(outputPath);
    
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    
    fs.writeFileSync(outputPath, rows.join('\n'));
    console.log(`\nTraining data generated: ${outputPath}`);
    console.log(`Total training samples: ${rows.length - 1}`);
    
    process.exit(0);
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

generateTrainingData();
